source/request_source/request_options_list_plugin_impl.cc

```diff
@@ -0,0 +1,93 @@
+
+namespace Nighthawk {
+std::string FileBasedRequestSourcePluginConfigFactory::name() const {
+  return "nighthawk.file-based-request-source-plugin";
+}
+
+Envoy::ProtobufTypes::MessagePtr
+FileBasedRequestSourcePluginConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<nighthawk::request_source::FileBasedPluginConfig>();
+}
+
+RequestSourcePtr FileBasedRequestSourcePluginConfigFactory::createRequestSourcePlugin(
+    const Envoy::Protobuf::Message& message, Envoy::Api::ApiPtr api,
+    Envoy::Http::RequestHeaderMapPtr header) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  nighthawk::request_source::FileBasedPluginConfig config;
+  Envoy::MessageUtil util;
+
+  util.unpackTo(any, config);
+  if (api->fileSystem().fileSize(config.file_path()) > config.max_file_size().value()) {
+    throw NighthawkException("file size must be less than max_file_size");
+  }
+  auto temp_list = std::make_unique<nighthawk::client::RequestOptionsList>();
+
+  // Locking to avoid issues with multiple threads reading the same file.
+  {
+    Envoy::Thread::LockGuard lock_guard(file_lock_);
+    // Reading the file only the first time.
+    if (options_list_.options_size() == 0) {
+      util.loadFromFile(config.file_path(), options_list_,
+                        Envoy::ProtobufMessage::getStrictValidationVisitor(), *api, true);
+    }
+    temp_list->CopyFrom(options_list_);
+  }
+  return std::make_unique<RequestOptionsListRequestSource>(config.num_requests().value(),
+                                                           std::move(header), std::move(temp_list));
+}
+
+REGISTER_FACTORY(FileBasedRequestSourcePluginConfigFactory, RequestSourcePluginConfigFactory);
+
+RequestOptionsListRequestSource::RequestOptionsListRequestSource(
+    const uint32_t request_max, Envoy::Http::RequestHeaderMapPtr header,
+    std::unique_ptr<nighthawk::client::RequestOptionsList> options_list)
+    : header_(std::move(header)), options_list_(std::move(options_list)),
+      request_max_(request_max) {}
+
+RequestGenerator RequestOptionsListRequestSource::get() {
+  uint32_t counter = 0;
+  request_count_.push_back(counter);
+  uint32_t& lambda_counter = request_count_.back();
+  RequestGenerator request_generator = [this, lambda_counter]() mutable -> RequestPtr {
+    // if request_max is 0, then we never stop generating requests.
+    if (lambda_counter >= request_max_ && request_max_ != 0) {
+      return nullptr;
+    }
+
+    // Increment the counter and get the request_option from the list for the current iteration.
+    auto index = lambda_counter % options_list_->options_size();
+    nighthawk::client::RequestOptions request_option = options_list_->options().at(index);
+    lambda_counter++;
+
+    // Initialize the header with the values from the default header.
+    Envoy::Http::RequestHeaderMapPtr header = Envoy::Http::RequestHeaderMapImpl::create();
+    Envoy::Http::HeaderMapImpl::copyFrom(*header, *header_);
```

Comment:
Since we are making a copy of the original header map anyway, why do we need to pass it in as a pointer to the plugin? Can't the plugin take the header as a const reference to make it clear on the API layer that the header isn't owned or modified?

Or is there a reason that forces us to give ownership?

Reply:
I think that it is more consistent with the way that the other plugins do it, and it is possible that the plugin would want to modify the header. Generally it also seems like something that the RequestSource should own. In the factories_impl.cc where this RequestSource will be created, I believe the header passed in does not have a lifetime greater than that of the RequestSource, and that might matter also, though I'm not sure if that is correct.